pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go

```diff
@@ -0,0 +1,88 @@
+package admitters
+
+import (
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	virtv1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+)
+
+type PodEvictionAdmitter struct {
+	ClusterConfig *virtconfig.ClusterConfig
+	VirtClient    kubecli.KubevirtClient
+}
+
+func (admitter *PodEvictionAdmitter) Admit(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	if !admitter.ClusterConfig.LiveMigrationEnabled() {
+		return &v1beta1.AdmissionResponse{Allowed: true}
+	}
+
+	launcher, err := admitter.VirtClient.CoreV1().Pods(ar.Request.Namespace).Get(ar.Request.Name, metav1.GetOptions{})
+	if err != nil {
+		// we don't want to block the request because we were not able to find the pod
+		return &v1beta1.AdmissionResponse{Allowed: true}
+	}
+
+	if value, exists := launcher.GetLabels()[virtv1.AppLabel]; !exists || value != "virt-launcher" {
+		// this is not a virt-launcher app so there's no reason to block it
+		return &v1beta1.AdmissionResponse{Allowed: true}
+	}
+
+	domainName, exists := launcher.GetAnnotations()[virtv1.DomainAnnotation]
+	if !exists {
+		// we don't know who is the VMI that created this pod so we will just have to allow this eviction
+		return &v1beta1.AdmissionResponse{Allowed: true}
+	}
+
+	vmi, err := admitter.VirtClient.VirtualMachineInstance(ar.Request.Namespace).Get(domainName, &metav1.GetOptions{})
+	if err != nil || !shouldTryInterceptEviction(vmi, launcher) {
+		return &v1beta1.AdmissionResponse{Allowed: true}
+	}
+
+	if vmi.MarkedForEviction() {
+		return makeRejectionResponse("virt-launcher eviction is already being handled by KubeVirt")
+	}
+
+	vmiCopy := vmi.DeepCopy()
+	vmiCopy.Status.EvictionState = &virtv1.VirtualMachineInstanceEvictionState{
+		MarkedForEvictionTimestamp: now(),
+		OriginalNodeName:           vmi.Status.NodeName,
+	}
+
+	_, err = admitter.VirtClient.VirtualMachineInstance(ar.Request.Namespace).Update(vmiCopy)
+	if err != nil {
+		// since we failed to mark the VMI, we must allow the eviction to proceed because there will be no retries
+		return &v1beta1.AdmissionResponse{Allowed: true}
```

Comment:
I'd expect the eviction to be retried by whatever is calling this if the POST request returns an error.  The client would have to be resilient to errors like this because they can already occur today if connectivity issues are encountered while issuing the evictions.

It would be safer for us to reject the eviction here and trust the client issuing the eviction to handle that. 

Our contract with the user when they set evictionStrategy to liveMigrate is that we're not going to stop/start the VM, but block until migration occurs.

Reply:
As we've discussed offline last week, I've moved this logic to a mutating webhook to ensure that we are being called. If this step will fail, the eviction will be blocked due to the PDB as expected.